npm/private/test/utils_tests.bzl

```diff
@@ -6,44 +6,85 @@ load("@bazel_skylib//lib:unittest.bzl", "asserts", "unittest")
+    asserts.equals(env, "x__meaning-of-life__1.0.0_124257499", utils.package_repo_name("x", "meaning-of-life@1.0.0(patch_hash=33610921243aecf4fa5a23dc8080659f436ccda15f41ce4f53c687039a305ee0)"))
+
+    # file:
     asserts.equals(env, "@scope+y@file+bar", utils.package_store_name("@scope/y@file:bar"))
-    asserts.equals(env, "@scope+y@file+_foo+bar", utils.package_store_name("@scope/y@file:@foo/bar"))
+    asserts.equals(env, "x__at_scope_y__file_bar", utils.package_repo_name("x", "@scope/y@file:bar"))
+
+    # file: + @
+    asserts.equals(env, "@scope+y@file+@foo+bar", utils.package_store_name("@scope/y@file:@foo/bar"))
+    asserts.equals(env, "x__at_scope_y__file__foo_bar", utils.package_repo_name("x", "@scope/y@file:@foo/bar"))
+
+    # file: ../
     asserts.equals(env, "@scope+y@file+..+foo+bar", utils.package_store_name("@scope/y@file:../foo/bar"))
+    asserts.equals(env, "x__at_scope_y__file_.._foo_bar", utils.package_repo_name("x", "@scope/y@file:../foo/bar"))
+
+    # file: ../tar
+    asserts.equals(env, "lodash@file+..+vendored+lodash-4.17.21.tgz", utils.package_store_name("lodash@file:../vendored/lodash-4.17.21.tgz"))
+    asserts.equals(env, "x__lodash__file_.._vendored_lodash-4.17.21.tgz", utils.package_repo_name("x", "lodash@file:../vendored/lodash-4.17.21.tgz"))
+
+    # file: .. (peers)
+    asserts.equals(env, "@scoped+c@file+..+projects+c_@scoped+b@projects+b", utils.package_store_name("@scoped/c@file:../projects/c(@scoped/b@projects+b)"))
+    asserts.equals(env, "x__at_scoped_c__file_.._projects_c_at_scoped_b_projects_b", utils.package_repo_name("x", "@scoped/c@file:../projects/c(@scoped/b@projects+b)"))
+
+    # URL
+    asserts.equals(env, "diff@https+++github.com+kpdecker+jsdiff+archive+refs+tags+v5.2.0.tar.gz", utils.package_store_name("diff@https://github.com/kpdecker/jsdiff/archive/refs/tags/v5.2.0.tar.gz"))
+    asserts.equals(env, "x__diff__https___github.com_kpdecker_jsdiff_archive_refs_tags_v5.2.0.tar.gz", utils.package_repo_name("x", "diff@https://github.com/kpdecker/jsdiff/archive/refs/tags/v5.2.0.tar.gz"))
+
+    # @ and 0.0.0 in a URL
+    asserts.equals(env, "@foo+jsonify@https+++github.com+aspect-build+test-packages+releases+download+0.0.0+@foo-jsonify-0.0.0.tgz", utils.package_store_name("@foo/jsonify@https://github.com/aspect-build/test-packages/releases/download/0.0.0/@foo-jsonify-0.0.0.tgz"))
+    asserts.equals(env, "x__at_foo_jsonify__https___github.com_aspect-build_test-packages_releases_download_0.0.0__foo-jsonify-0.0.0.tgz", utils.package_repo_name("x", "@foo/jsonify@https://github.com/aspect-build/test-packages/releases/download/0.0.0/@foo-jsonify-0.0.0.tgz"))
+
+    # URLs with commit hashes
+    asserts.equals(env, "jquery@https+++codeload.github.com+jquery+jquery+tar.gz+399b201bb3143a3952894cf3489b4848fc003967", utils.package_store_name("jquery@https://codeload.github.com/jquery/jquery/tar.gz/399b201bb3143a3952894cf3489b4848fc003967"))
```

Comment:
Note ugly ones like this are the exact same as pnpm 9, the `+++` and everything. The `file:../` ones converting to `file+..+` are also the exact same as pnpm